test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
-      sessionPool.reads_ = [{}];
-      sessionPool.writes_ = [{}, {}];
+    describe('createWriteSession_', function() {
+      it('should create a write session', function() {
+        const fakeSession = {
+          create: () => Promise.resolve(),
+        };
 
-      assert.strictEqual(sessionPool.available(), 3);
-    });
-  });
+        sessionPool.session_ = () => fakeSession;
 
-  describe('borrowed', function() {
-    it('should return the number of borrowed sessions', function() {
-      sessionPool.borrowed_ = [{}, {}];
+        sessionPool.createTransaction_ = function(session) {
+          assert.strictEqual(session, fakeSession);
+          return Promise.resolve();
+        };
 
-      assert.strictEqual(sessionPool.borrowed(), 2);
-    });
-  });
+        return sessionPool.createWriteSession().then(function(session) {
+          assert.strictEqual(session, fakeSession);
+          assert.strictEqual(session.type, 'readwrite');
+        });
+      });
 
-  describe('close', function() {
-    it('should set isOpen to false', function() {
-      sessionPool.isOpen = true;
-      sessionPool.close();
-      assert.strictEqual(sessionPool.isOpen, false);
-    });
+      it('should fail in creating a read/write session', function() {
+        let deleteCalled = false;
+        const fakeSession = {
+          create: () => Promise.reject('Error'),
+          delete: () => {
+            deleteCalled = true;
+            return Promise.resolve();
+          },
+        };
+
+        sessionPool.session_ = function() {
+          return fakeSession;
+        };
 
-    it('should emit a close event', function(done) {
-      sessionPool.on('close', done);
-      sessionPool.close();
+        return sessionPool.createWriteSession().catch(error => {
+          assert.strictEqual(error, 'Error');
+          assert.strictEqual(deleteCalled, true);
+        });
+      });
     });
 
-    it('should emit an empty event', function(done) {
-      sessionPool.on('empty', done);
-      sessionPool.close();
+    describe('open', function() {
+      it('should set isOpen to true', function() {
+        return sessionPool.close().then(function() {
+          sessionPool = new SessionPool(DATABASE, {minReads: 4, minWrites: 4});
+          sessionPool.session_ = function() {
+            return new Object({
+              create: () => Promise.resolve(),
+              delete: () => Promise.resolve(),
+            });
+          };
+          sessionPool.open();
+          assert.strictEqual(sessionPool.isOpen, true);
+          assert.strictEqual(sessionPool.getStats().readPool.size, 4);
+          assert.strictEqual(sessionPool.getStats().writePool.size, 0);
+          clearTimeout(sessionPool.pingTimeoutHandle);
+          return sessionPool.close();
+        });
+      });
     });
 
-    it('should call stopHouseKeeping_', function(done) {
-      sessionPool.stopHouseKeeping_ = done;
-      sessionPool.close();
+    describe('close', function() {
```

Comment:
There's a lot going on in this test, can we break it down into multiple test cases?

Reply:
This test case does the following:
1. open a session pool
2. check the readPool and writePool size
3. close the session pool
4. verify that the readPool and writePool size is 0

I don't think I can break down any more in this test case